pkg/controller/cloud/nodecontroller.go

```diff
@@ -90,10 +112,108 @@ func (cnc *CloudNodeController) Run() {
 	go func() {
 		defer utilruntime.HandleCrash()
 
+		instances, ok := cnc.cloud.Instances()
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
+			return
+		}
+
+		// Start a loop to periodically update the node addresses obtained from the cloud
 		go wait.Until(func() {
 			nodes, err := cnc.kubeClient.Core().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
 			if err != nil {
 				glog.Errorf("Error monitoring node status: %v", err)
+				return
+			}
+
+			for i := range nodes.Items {
+				node := &nodes.Items[i]
+
+				// Do not process nodes that are still tainted
+				taints := node.Spec.Taints
+
+				var cloudTaint *v1.Taint
+				for _, taint := range taints {
+					if taint.Key == CloudTaintKey {
+						cloudTaint = &taint
+					}
+				}
+
+				if cloudTaint != nil {
+					glog.V(5).Infof("This node %s is still tainted. Will not process.", node.Name)
+					continue
+				}
+
+				nodeAddresses, err := instances.NodeAddressesByProviderID(node.Spec.ProviderID)
+				if err != nil {
+					nodeAddresses, err = instances.NodeAddresses(types.NodeName(node.Name))
+					if err != nil {
+						glog.Errorf("failed to get node address from cloud provider: %v", err)
+						continue
+					}
+				}
+
+				var nodeIP net.IP
+				if ip, ok := node.ObjectMeta.Labels[LabelProvidedIPAddr]; ok {
+					nodeIP = net.ParseIP(ip)
+				}
+				// Check if a hostname address exists in the cloud provided addresses
+				hostnameExists := false
+				for i := range nodeAddresses {
+					if nodeAddresses[i].Type == v1.NodeHostName {
+						hostnameExists = true
+					}
+
+				}
+				// If hostname was not present in cloud provided addresses, use the hostname
+				// from the existing node (populated by kubelet)
+				var hostnameAddress *v1.NodeAddress
+				if !hostnameExists {
+					for _, addr := range node.Status.Addresses {
+						if addr.Type == v1.NodeHostName {
+							hostnameAddress = &addr
+						}
+					}
+				}
+				// If nodeIP was suggested by user, ensure that
+				// it can be found in the cloud as well (consistent with the behaviour in kubelet)
+				if nodeIP != nil {
+					var providedIP *v1.NodeAddress
+					for i := range nodeAddresses {
+						if nodeAddresses[i].Address == nodeIP.String() {
+							providedIP = &nodeAddresses[i]
+						}
+					}
+					if providedIP == nil {
+						glog.Errorf("failed to get node address from cloudprovider that matches ip: %v", nodeIP)
+						continue
+					}
+					nodeAddresses = []v1.NodeAddress{
```

Comment:
This is exactly what kubelet does now - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L407